### DISCUSSION TITLE
Delete role_has_permissions table

### DISCUSSION THREAD

**Discussion by Mahan-Shoghy** [spatie/laravel-permission#2102] [Ideas]
Hello!
Why did you keep this table? Use model_has_permissions instead?

**Comment by erikn69** [comment]
Because roles can have permissions

**Reply by Mahan-Shoghy** [reply]
Of course! But I mean model_has_permissions is polymorphic relation and can be used for this purpose,
Why we have to create another table for this?

**Reply by erikn69** [reply]
Because me and many others use Role_has_permissions instead of model_has_permissions, maybe is better if spatie  delete model_has_permission, did you see the problem? Big breaking change

**Comment by Mahan-Shoghy** [comment]
Spatie\Permission\Models\Role class use HasPermissions trait
![image](https://user-images.githubusercontent.com/55574901/169888385-1380835f-afb4-467c-b4bc-d4ed946b5a7a.png)

in Spatie\Permission\Traits\HasPermissions we have permissions() that return polymorphic relation with model_has_permissions table
![image](https://user-images.githubusercontent.com/55574901/169888746-961cf6c1-da0b-4082-9b97-b18b9cb559fd.png)

again in Spatie\Permission\Models\Role class we rewrite permissions() for return many to many relation
![image](https://user-images.githubusercontent.com/55574901/169889452-11976047-b45d-484d-9b5d-177ee45825a0.png)

I mean, the rewrite method is extra and can be removed, because roles can have permissions with model_has_permissions table too.
my question is why role_has_permissions table exists? for more readability? or is exists to allow developers for create custom table? I want to know

**Comment by erikn69** [comment]
If you reuse the same table for those relationships, it breaks teams feature, also it would be a catastrophic breaking change for everyone
https://github.com/spatie/laravel-permission/blob/8a69aaddbf91810b0175b05a1acb9b3410bff68e/database/migrations/create_permission_tables.php.stub#L68-L69
Also if you have custom pivot names everything would get worse
https://github.com/spatie/laravel-permission/blob/8a69aaddbf91810b0175b05a1acb9b3410bff68e/config/permission.php#L78
https://github.com/spatie/laravel-permission/blob/8a69aaddbf91810b0175b05a1acb9b3410bff68e/config/permission.php#L89
Why the desire to complicate things for everyone?

**Reply by Mahan-Shoghy** [reply]
Oh thank you for your answer. This is not to complicate things, I're just challenging to learn, and it has always been my question to know why this happened. After looking at the bouncer package and seeing the tables are  not separate, I dared to ask.

Anyway, I very thank you for your answer

**Reply by erikn69** [reply]
Mark this as answered
